### PR TITLE
Add GestureMonitor and hooks

### PR DESCRIPTION
**Pull request by j-piasecki**
## Description

Add `GestureMonitor`, `useGesture` and `useAnimatedGesture`, which allows for using the new API.

## Test plan

Added a new example to the Example app and tested on it

### CODE REVIEW

File: src/handlers/gestures/useGesture.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  function needsToReattach() {
+    if (gesture.length !== result.current.config.length) {
+      return true;
+    } else {
+      for (let i = 0; i < gesture.length; i++) {
+        if (gesture[i].handlerName !== result.current.config[i].handlerName) {
+          return true;
+        }
+      }
+    }
```
Comment (jakub-gonet) [inline src/handlers/gestures/useGesture.ts:192]: You don't need `else` here.

File: src/handlers/gestures/useGesture.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    for (let i = 0; i < gesture.length; i++) {
+      const gst = result.current.config[i];
+
+      gesture[i].handlerTag = gst.handlerTag;
+      gesture[i].handlers.handlerTag = gst.handlerTag;
+    }
+
+    for (let i = 0; i < gesture.length; i++) {
+      const gst = result.current.config[i];
```
Comment (jakub-gonet) [inline src/handlers/gestures/useGesture.ts:145]: Please rename it.

File: src/handlers/gestures/useGesture.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+          .map(convertToHandlerTag)
+          .filter((tag) => tag > 0);
+      }
+
+      let simultaneousWith: number[] = [];
+      if (gst.config.simultaneousWith) {
+        simultaneousWith = gst.config.simultaneousWith
+          .map(convertToHandlerTag)
+          .filter((tag) => tag > 0);
+      }
```
Comment (jakub-gonet) [inline src/handlers/gestures/useGesture.ts:163]: 💡 

```suggestion
      const simultaneousWith: number[] =
      gst.config.simultaneousWith
          ?.map(convertToHandlerTag)
          ?.filter((tag) => tag > 0) ?? [];
```

File: src/handlers/gestures/useGesture.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        (g) => g.handlers
+      ) as unknown) as HandlerCallbacks<Record<string, unknown>>[];
+    }
+  }
+
+  function updateHandlers() {
+    gestureConfig.prepare();
+
+    for (let i = 0; i < gesture.length; i++) {
+      const gst = result.current.config[i];
```
Comment (jakub-gonet) [inline src/handlers/gestures/useGesture.ts:138]: Please rename it.

File: src/handlers/gestures/useGesture.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  function attachHandlers() {
+    if (!result.current.firstExecution) {
+      gestureConfig.initialize();
+    } else {
+      result.current.firstExecution = false;
+    }
+
+    for (const gst of gesture) {
```
Comment (jakub-gonet) [inline src/handlers/gestures/useGesture.ts:89]: Please rename.

File: src/handlers/gestures/useGesture.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+export function useGesture(gestureConfig: InteractionBuilder | GestureType) {
+  const gesture = gestureConfig.configure();
+
+  const result = React.useRef<GestureConfigReference>({
+    config: gesture,
+    callback: null,
+    animatedEventHandler: null,
+    animatedHandlers: null,
+    firstExecution: true,
+  });
```
Comment (jakub-gonet) [inline src/handlers/gestures/useGesture.ts:54]: You can add `.current` here to decrease verbosity.

**Review by jakub-gonet — CHANGES_REQUESTED** [review]

File: examples/Example/src/new_api/index.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+function Draggable() {
+  const [counter, setCounter] = useState(1);
+
+  useEffect(() => {
+    setInterval(() => {
+      setCounter((a) => a + 1);
+    }, 1000);
+  }, []);
+
+  const pressed = useSharedValue(false);
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/index.tsx]: Fixed in 13a9934a344b400afa5257d2a1bc60f3907c9bb6

File: examples/Example/src/new_api/index.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      transform: [
+        { translateX: offset.value.x },
+        { translateY: offset.value.y },
+        { scale: withSpring(pressed.value ? 1.2 : 1) },
+      ],
+      backgroundColor: pressed.value ? 'yellow' : 'blue',
+    };
+  });
+
+  const gs = useAnimatedGesture(
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/index.tsx]: Fixed in 13a9934a344b400afa5257d2a1bc60f3907c9bb6

File: examples/Example/src/new_api/index.tsx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    Gesture.pan()
+      .setOnBegan((_e) => {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/index.tsx]: Removed them in 765e00c2752e68e3b380fffd196d84b1a09a6205

File: examples/Example/src/new_api/index.tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  return (
+    <Animated.View>
+      <GestureMonitor gesture={gs}>
+        <Element styles={animatedStyles} counter={counter} />
+      </GestureMonitor>
+    </Animated.View>
+  );
+}
+
+function Element(props: { styles: unknown; counter: number }) {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/index.tsx]: Unfortunately it's not `CSSProperties` as those styles are from `useAnimatedStyle` but I updated the type in 448ff5085779d302259d510eaec835ee0eeee34e.

File: examples/Example/src/new_api/index.tsx [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+export default function Example() {
+  return (
+    <View style={styles.home}>
+      <Draggable />
+    </View>
+  );
+}
+
+const styles = StyleSheet.create({
+  home: {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/index.tsx]: Fixed in 13a9934a344b400afa5257d2a1bc60f3907c9bb6

File: src/handlers/gestureHandlerCommon.ts [hunk 12]
```diff
@@ -29,6 +31,8 @@ export const baseGestureHandlerProps = [
   'onHandlerStateChange',
 ] as const;
 
+export const baseGestureHandlerWithMonitorProps = [...commonProps] as const;
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestureHandlerCommon.ts]: Fixed in 5b8b13b068a9503096e82199faea7b67a02d24a9

File: src/handlers/gestureHandlerCommon.ts [hunk 13]
```diff
@@ -102,3 +106,57 @@ export type BaseGestureHandlerProps<
...
+  );
+}
+
+export function filterConfig(
+  props: Record<string, unknown>,
+  validProps: string[],
+  defaults: Record<string, unknown> = {}
+) {
+  const res = { ...defaults };
+  validProps.forEach((key) => {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestureHandlerCommon.ts]: Changed in 25ccfe77bf72889dc9ebad230c6e24b5d7b9af44

File: src/handlers/gestureHandlerCommon.ts [hunk 14]
```diff
@@ -102,3 +106,57 @@ export type BaseGestureHandlerProps<
...
+    name !== 'onGestureEvent'
+  );
+}
+
+export function filterConfig(
+  props: Record<string, unknown>,
+  validProps: string[],
+  defaults: Record<string, unknown> = {}
+) {
+  const res = { ...defaults };
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestureHandlerCommon.ts]: Fixed in 25ccfe77bf72889dc9ebad230c6e24b5d7b9af44

File: src/handlers/gestureHandlerCommon.ts [hunk 15]
```diff
@@ -102,3 +106,57 @@ export type BaseGestureHandlerProps<
...
+  validProps.forEach((key) => {
+    const value = props[key];
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestureHandlerCommon.ts]: Changed in 25ccfe77bf72889dc9ebad230c6e24b5d7b9af44

File: src/handlers/gestureHandlerCommon.ts [hunk 16]
```diff
@@ -102,3 +106,57 @@ export type BaseGestureHandlerProps<
...
+    if (isConfigParam(value, key)) {
+      let value = props[key];
+      if (key === 'simultaneousHandlers' || key === 'waitFor') {
+        value = transformIntoHandlerTags(props[key]);
+      } else if (key === 'hitSlop') {
+        if (typeof value !== 'object') {
+          value = { top: value, left: value, bottom: value, right: value };
+        }
+      }
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestureHandlerCommon.ts]: Changed in 25ccfe77bf72889dc9ebad230c6e24b5d7b9af44

File: src/handlers/gestureHandlerCommon.ts [hunk 17]
```diff
@@ -102,3 +106,57 @@ export type BaseGestureHandlerProps<
...
+      res[key] = value;
+    }
+  });
+  return res;
+}
+
+function transformIntoHandlerTags(handlerIDs: any) {
+  if (!Array.isArray(handlerIDs)) {
+    handlerIDs = [handlerIDs];
+  }
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestureHandlerCommon.ts]: Done in 0d595327e5c26f62f6d1a5cc94d1b4105833676e

File: src/handlers/gestures/gesture.ts [hunk 18]
```diff
@@ -1,30 +1,36 @@
...
-    | React.RefObject<BaseGesture<Record<string, unknown>>>
-  )[];
-  simultaneousWith?: (
-    | number
-    | BaseGesture<Record<string, unknown>>
-    | React.RefObject<BaseGesture<Record<string, unknown>>>
-  )[];
+  ref?: React.MutableRefObject<GestureType>;
+  requireToFail?: (number | GestureType | React.RefObject<GestureType>)[];
+  simultaneousWith?: (number | GestureType | React.RefObject<GestureType>)[];
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/gesture.ts]: Fixed in fe30cf16ef8c184f2d0a22df7e7a9fb05caaf80f

File: src/handlers/gestures/useAnimatedGesture.ts [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    if (!currentCallback) {
+      return;
+    }
+
+    //eslint-disable-next-line @typescript-eslint/prefer-for-of
+    for (let i = 0; i < currentCallback.length; i++) {
+      const gesture = currentCallback[i];
+
+      if (event.handlerTag === gesture.handlerTag) {
+        const e = event as UnwrappedGestureHandlerStateChangeEvent;
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useAnimatedGesture.ts]: Done in 2691768467b88e13cb14d658bf93a55d355892fd

File: src/handlers/gestures/useAnimatedGesture.ts [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        if (e.oldState != null) {
+          if (e.oldState === State.UNDETERMINED && e.state === State.BEGAN) {
+            gesture.onBegan?.(e);
+          } else if (
+            (e.oldState === State.BEGAN || e.oldState === State.UNDETERMINED) &&
+            e.state === 4
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useAnimatedGesture.ts]: Done in 2691768467b88e13cb14d658bf93a55d355892fd

File: src/handlers/gestures/useGesture.ts [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+
+  if (result.current.firstExecution) {
+    gestureConfig.initialize();
+  }
+
+  function dropHandlers() {
+    for (const g of result.current.config) {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in a2718399f38db831195c44bd8645ff26ad4c25ff

File: src/handlers/gestures/useGesture.ts [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+  function convertToHandlerTag(ref: GestureRef): number {
+    if (typeof ref === 'number') {
+      return ref;
+    } else if (ref instanceof BaseGesture) {
+      return ref.handlerTag;
+    } else {
+      let tag = ref.current?.handlerTag;
+      if (tag === undefined) {
+        tag = -1;
+      }
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in c4e68c41a9208dafa2dbb5e6f976bfabaeb92674

File: src/handlers/gestures/useGesture.ts (hunk 1 above)
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Removed in 0acba7d437cd505513529f02eeca5d5610bba6b5

File: src/handlers/gestures/useGesture.ts [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+  animatedHandlers: Animated.SharedValue<
+    HandlerCallbacks<Record<string, unknown>>[] | null
+  > | null;
+  firstExecution: boolean;
+};
+
+export function useGesture(gestureConfig: InteractionBuilder | GestureType) {
+  const gesture = gestureConfig.configure();
+
+  const result = React.useRef<GestureConfigReference>({
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Renamed it to `preparedGesture` in 6a45f245c165bf9fc2a2bddbebe5019406a2a44d. Is that name ok?

File: src/handlers/gestures/useGesture.ts (hunk 2 above)
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in 61476ebafddc567f70dbfce5e0a55664510af638

File: src/handlers/gestures/useGesture.ts (hunk 4 above)
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in 61476ebafddc567f70dbfce5e0a55664510af638

File: src/handlers/gestures/useGesture.ts (hunk 6 above)
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in 991221b8434e0777733a763c092c962fe2d368e9

File: src/handlers/gestures/useGesture.ts (hunk 5 above)
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in b1535912f84e63f33bf3e83e92be2ebea6a5de84

File: src/handlers/gestures/useGesture.ts (hunk 3 above)
Comment (j-piasecki) [reply src/handlers/gestures/useGesture.ts]: Done in 8602322bfdd6bcf87f63654c882c2a80814a9fbe

File: src/handlers/gestures/useAnimatedGesture.ts [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+  ) => {
+    'worklet';
+
+    const currentCallback = sharedHandlersCallbacks.value;
+    if (!currentCallback) {
+      return;
+    }
+
+    //eslint-disable-next-line @typescript-eslint/prefer-for-of
+    for (let i = 0; i < currentCallback.length; i++) {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useAnimatedGesture.ts:38]: Apparently it causes problems with `worklets`, when I tried it there the app would crash.

File: src/handlers/gestures/eventReceiver.ts [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+  event: UnwrappedGestureHandlerStateChangeEvent
+) {
+  const gesture = findHandler(event.handlerTag);
+
+  if (gesture) {
+    if (event.oldState === State.UNDETERMINED && event.state === State.BEGAN) {
+      gesture.handlers.onBegan?.(event);
+    } else if (
+      (event.oldState === State.BEGAN ||
+        event.oldState === State.UNDETERMINED) &&
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/eventReceiver.ts:30]: I don't know about this one. Imo it doesn't improve readability that much and is inconsistent with other `if` branches. (And TypeScript yells at me about incompatible types).

File: src/handlers/gestures/useAnimatedGesture.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/handlers/gestures/useAnimatedGesture.ts:38]: Could you add that in or above the eslint comment to give more context in the future (and possibly change it when Reanimated starts supporting for-ofs).

File: src/handlers/gestures/eventReceiver.ts [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+      event.state === State.ACTIVE
+    ) {
+      gesture.handlers.onStart?.(event);
+    } else if (event.oldState === State.ACTIVE && event.state === State.END) {
+      gesture.handlers.onEnd?.(event, true);
+    } else if (event.state === State.FAILED) {
+      gesture.handlers.onEnd?.(event, false);
+    } else if (event.state === State.CANCELLED) {
+      gesture.handlers.onEnd?.(event, false);
+    }
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/eventReceiver.ts]: Same as above, but I have merged the branches with the same body in e3d40985c517c1e2b1d3eabdf32ef7203b044804.

File: src/handlers/gestures/useGesture.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/handlers/gestures/useGesture.ts]: Seems alright.

File: src/handlers/gestures/useAnimatedGesture.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/handlers/gestures/useAnimatedGesture.ts:40]: Done in 81f4bd3c5d8813c02198dbbf7d735c0e909b2338

File: src/handlers/gestures/GestureMonitor.tsx [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+          {this.props.children}
+        </AnimatedWrap>
+      );
+    } else {
+      return React.Children.only(this.props.children);
+    }
+  }
+}
+
+class Wrap extends React.Component<{ onGestureHandlerEvent: any }> {
```
Comment (jakub-gonet) [inline src/handlers/gestures/GestureMonitor.tsx:91]: `any`

File: src/handlers/gestures/eventReceiver.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/handlers/gestures/eventReceiver.ts:30]: Sure, let's leave it then.

File: examples/Example/src/new_api/index.tsx [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+        translateY: number;
+        translateX?: undefined;
+        scale?: undefined;
+      }
+    | {
+        scale: number;
+        translateX?: undefined;
+        translateY?: undefined;
+      }
+  )[];
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/new_api/index.tsx]: Done in b6df51f48d1907ee8759e006b0cc77dd0c530e50

File: src/handlers/gestures/GestureMonitor.tsx (hunk 27 above)
Comment (j-piasecki) [reply src/handlers/gestures/GestureMonitor.tsx]: Changed it to `unknown` in b6df51f48d1907ee8759e006b0cc77dd0c530e50

**Review by jakub-gonet — APPROVED** [review]

**Pull request merged**